### PR TITLE
CodeBlock: Use static memory for JIT locations

### PR DESCRIPTION
**Pull request by degasus**
This PR uses a global array for the code region of the DSP LLE emitter. The idea is to get rid of the "AllocateExecutableMemory" function, which has shown to be an issue with PIE binaries.

### CODE REVIEW

**Comment by degasus** [comment]
@grp Do you think this is a good idea with iOS in mind?

File: Source/Core/Common/MemoryUtil.cpp [hunk 1]
```diff
@@ -165,62 +165,54 @@ void FreeAlignedMemory(void* ptr)
...
-
-void WriteProtectMemory(void* ptr, size_t size, bool allowExecute)
-{
-  bool error_occurred = false;
-
 #ifdef _WIN32
+  DWORD newValue;
   DWORD oldValue;
-  if (!VirtualProtect(ptr, size, allowExecute ? PAGE_EXECUTE_READ : PAGE_READONLY, &oldValue))
+  switch ((read << 0) | (write << 1) | (exec << 2))
```
Comment (lioncash) [inline Source/Core/Common/MemoryUtil.cpp]: This switch should probably be its own function.

File: Source/Core/Common/MemoryUtil.h [hunk 2]
```diff
@@ -14,10 +14,22 @@ void* AllocateMemoryPages(size_t size);
...
 void FreeAlignedMemory(void* ptr);
-void ReadProtectMemory(void* ptr, size_t size);
-void WriteProtectMemory(void* ptr, size_t size, bool executable = false);
-void UnWriteProtectMemory(void* ptr, size_t size, bool allowExecute = false);
+bool MemProtect(void* ptr, size_t size, bool read, bool write, bool exec);
 std::string MemUsage();
 size_t MemPhysical();
 
+// Deprecated
+inline void ReadProtectMemory(void* ptr, size_t size)
```
Comment (lioncash) [inline Source/Core/Common/MemoryUtil.h]: These can have their definitions placed in the cpp file.

File: Source/Core/Common/CodeBlock.h [hunk 3]
```diff
@@ -38,11 +38,21 @@ class CodeBlock : public T, NonCopyable
       FreeCodeSpace();
   }
 
+  void AllocCodeSpace(int size, bool need_low)
```
Comment (lioncash) [inline Source/Core/Common/CodeBlock.h]: `size_t size`

File: Source/Core/Common/CodeBlock.h [hunk 4]
```diff
@@ -38,11 +38,21 @@ class CodeBlock : public T, NonCopyable
...
+    if (need_low)
+      region = static_cast<u8*>(Common::AllocateExecutableMemory(size, need_low));
+    else
+      region = static_cast<u8*>(Common::AllocateMemoryPages(size));
+    SetCodeSpace(region, size);
+  }
+
   // Call this before you generate any code.
-  void AllocCodeSpace(int size, bool need_low = true)
+  void SetCodeSpace(u8* _region, int size)
```
Comment (lioncash) [inline Source/Core/Common/CodeBlock.h]: `region_` also `size_t size`

File: Source/Core/Core/DSP/DSPEmitter.cpp [hunk 5]
```diff
@@ -16,11 +16,13 @@
 
 using namespace Gen;
 
+alignas(4096) u8 code_region[COMPILED_CODE_SIZE];
```
Comment (lioncash) [inline Source/Core/Core/DSP/DSPEmitter.cpp]: std::array. This should also probably be a member of the DSPEmitter.

**Comment by grp** [comment]
@degasus For iOS the ideal setup is writing out any executable code once, then marking it read-execute after that. That of course doesn't work with (at least) back patching, so second best is a hook before/after the writes. AFAICT this doesn't change that at all, so would just need a rebase.

One thing that would be a tiny bit harder is the grep command to find places marking memory writable. Rather than three booleans, maybe a bitfield with constants to search for?

File: Source/Core/Core/DSP/DSPEmitter.h [hunk 6]
```diff
@@ -263,6 +264,8 @@ class DSPEmitter : public Gen::X64CodeBlock
   // Counts down.
   // int cycles;
 
+  static std::array<u8, COMPILED_CODE_SIZE> code_region;
```
Comment (lioncash) [inline Source/Core/Core/DSP/DSPEmitter.h]: Why would this need to be static? Hypothetically, if you ran two DSPEmitter instances at once, you'd want them to have their own code regions, no?.

File: Source/Core/Core/PowerPC/Jit64/Jit.h [hunk 7]
```diff
@@ -33,6 +35,8 @@ class Jit64 : public Jitx86Base
   void AllocStack();
   void FreeStack();
 
+  static std::array<u8, CODE_SIZE> code_area;
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/Jit64/Jit.h]: Ditto

File: Source/Core/Core/PowerPC/JitCommon/Jit_Util.h [hunk 8]
```diff
@@ -27,38 +28,41 @@ class Mapping;
...
+// When making large additions to the AsmCommon code, you might want to ensure this number is big
+// enough.
+static const size_t ASM_CODE_SIZE = 1024 * 16;
+
 // A place to throw blocks of code we don't want polluting the cache, e.g. rarely taken
 // exception branches.
 class FarCodeCache : public Gen::X64CodeBlock
 {
 private:
+  static std::array<u8, FARCODE_SIZE> code_area;
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitCommon/Jit_Util.h]: Ditto

File: Source/Core/Core/PowerPC/JitCommon/TrampolineCache.h [hunk 9]
```diff
@@ -19,9 +21,10 @@ class TrampolineCache : public EmuCodeBlock
   const u8* GenerateReadTrampoline(const TrampolineInfo& info);
   const u8* GenerateWriteTrampoline(const TrampolineInfo& info);
 
+  static std::array<u8, TRAMPOLINE_CODE_SIZE> code_area;
```
Comment (lioncash) [inline Source/Core/Core/PowerPC/JitCommon/TrampolineCache.h]: Ditto

File: Source/Core/Common/MemoryUtil.h [hunk 10]
```diff
@@ -14,10 +14,13 @@ void* AllocateMemoryPages(size_t size);
...
 void FreeAlignedMemory(void* ptr);
-void ReadProtectMemory(void* ptr, size_t size);
-void WriteProtectMemory(void* ptr, size_t size, bool executable = false);
-void UnWriteProtectMemory(void* ptr, size_t size, bool allowExecute = false);
+bool MemProtect(void* ptr, size_t size, bool read, bool write, bool exec);
 std::string MemUsage();
 size_t MemPhysical();
 
+// Deprecated
+void ReadProtectMemory(void* ptr, size_t size);
```
Comment (lioncash) [inline Source/Core/Common/MemoryUtil.h]: Just an idea, but C++ has the `[[deprecated]]` attribute that would be used like:

```cpp
[[deprecated("newer_func() should be used instead")]]
void some_func();
```

File: Source/Core/Core/DSP/DSPEmitter.h (hunk 6 above)
Comment (degasus) [inline Source/Core/Core/DSP/DSPEmitter.h]: Yes, but they *have* to be stored close to the C++ code, not in the heap.

File: Source/Core/Common/MemoryUtil.h (hunk 10 above)
Comment (degasus) [inline Source/Core/Common/MemoryUtil.h]: Quite long messages. I think it's better to just fix the callers ;)

File: Source/Core/Core/PowerPC/Jit64/Jit.cpp [hunk 11]
```diff
@@ -39,6 +39,8 @@
 using namespace Gen;
 using namespace PowerPC;
 
+alignas(4096) std::array<u8, CODE_SIZE> Jit64::code_area;
```
Comment (hthh) [inline Source/Core/Core/PowerPC/Jit64/Jit.cpp]: Could we put the constant "4096" in one place?

I ask because the page size on iOS in some circumstances is 16kb (not sure if it matters for Dolphin), but it is something they changed recently.

I also ask because x86 can sometimes use 2mb/4mb page sizes, and I think there's a (very small) chance the slowdown is caused by forcing the use of 4kb pages, and stressing the TLB. Changing the alignment to 4mb would be a good test to rule out that theory.

File: Source/Core/Core/DSP/DSPEmitter.cpp [hunk 12]
```diff
@@ -16,11 +16,13 @@
 
 using namespace Gen;
 
+alignas(4096) std::array<u8, COMPILED_CODE_SIZE> DSPEmitter::code_region;
```
Comment (hthh) [inline Source/Core/Core/DSP/DSPEmitter.cpp]: I'd probably call this `code_area` to match the others

**Comment by degasus** [comment]
JMC47 |  it's actually quite a bit faster than master now

I wonder why this affects performance at all, but a bit performance noise may be fine.

@hthh Fixed.

@lioncash So may I add those static buffers? I also dislike them, but I see no way without a big redesign and a slowdown in the x64 JITs.

**Comment by lioncash** [comment]
@degasus I've explained at length why I disagree with it on IRC 

tl;dr version for others: it is a step back with regards to instance flexibility, *especially* in the context of libdolphin. It's "good" (finger-quotes) for short-term gain, but bad for maintainability in the long-term. It also detracts from amending the JIT itself by adding more work to undo in that regard.

I'm sorry, but I'm not willing to back a change like that.

**Comment by degasus** [comment]
@lioncash libdolphin is one of my main reasons to write this PR. This PR makes it explicit that the x64 JITs can only run once per process. This is IMO by far better than requireing that less than 2G of memory are allocated while booting the JIT. The latter requirement is the main reason why we can't switch the CPU emulation (interpreter <-> JIT) at runtime.

But indeed. Dropping the 2G of allocated memory requirement will remove the pressure to change the JIT to use 64bit pointers. But it won't make it harder.

Note: Using 64bit pointers will be a slowdown as it increases the code size quite a bit. But likely only a small slowdown.

File: Source/Core/Common/MemoryUtil.h [hunk 13]
```diff
@@ -9,15 +9,11 @@
...
 {
-void* AllocateExecutableMemory(size_t size, bool low = true);
 void* AllocateMemoryPages(size_t size);
 void FreeMemoryPages(void* ptr, size_t size);
 void* AllocateAlignedMemory(size_t size, size_t alignment);
 void FreeAlignedMemory(void* ptr);
-void ReadProtectMemory(void* ptr, size_t size);
-void WriteProtectMemory(void* ptr, size_t size, bool executable = false);
-void UnWriteProtectMemory(void* ptr, size_t size, bool allowExecute = false);
+bool MemProtect(void* ptr, size_t size, bool read, bool write, bool exec);
```
Comment (delroth) [inline Source/Core/Common/MemoryUtil.h]: Having 3 booleans is completely unreadable when looking at call sites. Make this a set of flags.

File: Source/Core/Common/CodeBlock.h [hunk 14]
```diff
@@ -8,6 +8,8 @@
 #include "Common/MemoryUtil.h"
 #include "Common/NonCopyable.h"
 
+constexpr size_t JIT_MEM_ALIGNMENT = 4096;
```
Comment (delroth) [inline Source/Core/Common/CodeBlock.h]: Comment on what this constant is about and why it has this value.

File: Source/Core/Core/PowerPC/Jit64/Jit.cpp [hunk 15]
```diff
@@ -39,6 +39,8 @@
 using namespace Gen;
 using namespace PowerPC;
 
+alignas(JIT_MEM_ALIGNMENT) std::array<u8, CODE_SIZE> Jit64::code_area;
```
Comment (delroth) [inline Source/Core/Core/PowerPC/Jit64/Jit.cpp]: Comment on the use of this global (either here or in its declaration).

**Comment by delroth** [comment]
Have you measured the impact in statup time of having to clear ~150MB of RAM in .bss?

File: Source/Core/Core/PowerPC/Jit64/Jit.cpp (hunk 15 above)
Comment (degasus) [inline Source/Core/Core/PowerPC/Jit64/Jit.cpp]: I've done this in the header.

**Comment by degasus** [comment]
I've dropped the first commit with the MemProtect helper as @grp and @delroth seem to dislike such a combined API.

@delroth No, I haven't meassured it. But I'd expect it to be bottlenecked at DDR 15 GB/s speed, so just about 10 ms?

**Comment by degasus** [comment]
#5271 is the better one for now.
